server/src/main/java/org/elasticsearch/common/io/stream/NetworkStreamOutput.java

```diff
@@ -0,0 +1,227 @@
+            return;
+        }
+
+        // illegal args: offset and/or length exceed array size
+        if (b.length < (offset + length)) {
+            throw new IllegalArgumentException("Illegal offset " + offset + "/length " + length + " for byte[] of length " + b.length);
+        }
+
+        // get enough pages for new size
+        ensureCapacity(length);
+
+        // bulk copy
+        int bytesToCopy = length;
+        int srcOff = offset;
+        int j = 0;
+        while (true) {
+            BytesRef currentPage = pages.get(pageIndex + j).v();
+            int toCopyThisLoop = Math.min(pageSize - currentPageOffset, bytesToCopy);
+            System.arraycopy(b, srcOff, currentPage.bytes, currentPage.offset + currentPageOffset, toCopyThisLoop);
+            srcOff += toCopyThisLoop;
+            bytesToCopy -= toCopyThisLoop;
+            if (bytesToCopy > 0) {
+                currentPageOffset = 0;
+            } else {
+                currentPageOffset += toCopyThisLoop;
+                break;
+            }
+            j++;
+        }
+
+        // advance
+        pageIndex += j;
+    }
+
+    @Override
+    public void writeInt(int i) throws IOException {
+        if (4 > (pageSize - currentPageOffset)) {
+            super.writeInt(i);
+        } else {
+            BytesRef currentPage = pages.get(pageIndex).v();
+            VH_BE_INT.set(currentPage.bytes, currentPage.offset + currentPageOffset, i);
+            currentPageOffset += 4;
+        }
+    }
+
+    @Override
+    public void writeLong(long i) throws IOException {
+        if (8 > (pageSize - currentPageOffset)) {
+            super.writeLong(i);
+        } else {
+            BytesRef currentPage = pages.get(pageIndex).v();
+            VH_BE_LONG.set(currentPage.bytes, currentPage.offset + currentPageOffset, i);
+            currentPageOffset += 8;
+        }
+    }
+
+    @Override
+    public void reset() {
+        for (Recycler.V<BytesRef> page : pages) {
+            page.close();
+        }
+        pages.clear();
+        pageIndex = -1;
+        currentPageOffset = pageSize;
+    }
+
+    @Override
+    public void flush() {
+        // nothing to do
+    }
+
+    @Override
+    public void seek(long position) {
+        ensureCapacityFromPosition(position);
+        this.pageIndex = (int) position / pageSize;
+        this.currentPageOffset = (int) position % pageSize;
+    }
+
+    public void skip(int length) {
+        seek(position() + length);
+    }
+
+    @Override
+    public void close() {
+        for (Recycler.V<BytesRef> page : pages) {
+            page.close();
+        }
+        pages.clear();
+    }
+
+    /**
+     * Returns the current size of the buffer.
+     *
+     * @return the value of the <code>count</code> field, which is the number of valid
+     *         bytes in this output stream.
+     * @see ByteArrayOutputStream#size()
+     */
+    public int size() {
+        return (int) position();
```

Comment:
This cast is supposed to be safe given the check in ensureCapacity, but maybe still use `Math.toIntExact` for safety?